File: 84-largest-rectangle-in-histogram/largest-rectangle-in-histogram.cpp
class Solution {
public:
    int largestRectangleArea(vector<int>& heights) {
        int n = heights.size();
        int width = 0;
        int maxarea = 0;
        stack<int> st;
        for(int i=0; i<=n; i++){
            while(!st.empty() && (i == n || heights[st.top()] >= heights[i])){
                int h = heights[st.top()];
                st.pop();
                if(st.empty()){
                    width = i;
                }
                else{
                    width = i - st.top() - 1;
                }
                maxarea = max(maxarea, h * width);
            }
            st.push(i);
        }
        return maxarea;
    }
};

File: 84-largest-rectangle-in-histogram/largest-rectangle-in-histogram.py
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        heights.append(0)

        st = []
        maxarea = 0
        for i in range(len(heights)):
            while st and heights[st[-1]] >= heights[i]:
                h = heights[st.pop()]
                if not st:
                    width = i
                else:
                    width = i - st[-1] - 1
                maxarea = max(maxarea, h * width)
            st.append(i)
        return maxarea